pkg/client/restclient/urlprovider.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restclient
+
+import (
+	"net/url"
+	"sync"
+)
+
+// A URLProvider allows to use multiple URLs.
+// Also maintains state of currently working URL.
+type URLProvider interface {
+	// Get returns currently selected URL.
+	Get() *url.URL
+	// Next selects any other URL, it is upto implementation to provide any
+	// health checking/load balancing, it might be as simple as an iterator.
+	Next() *url.URL
+	// WrapWithRetry executes function until success or until all urls were visited.
+	WrapWithRetry(func() error) error
+}
+
+// A RoundRobinProvider allows to iterate over URLs, doesnt provide any health checking.
+type RoundRobinProvider struct {
```

Comment:
1. In current implementation switch will happen only on error, e.g. there is no internal load balancer, and we will stick to the first url as long as it will work.
Sure, Next is public, but it is used only for testing.

2. There is 2 differences with https://github.com/dshulyak/kubernetes/blob/c628b0f279a2581b3b59d43751289cf66740c65a/pkg/client/restclient/slightly_sticky_provider.go, in slightly sticky provider implementation:

   - URL selected based on time
   - switch happens not on **first** error, but certain burst of errors can be tolerated

    Why those things will provide stronger session affinity guarantees?

3. @smarterclayton you are probably right about interface. If we will want to provide such configuration as errors qps and errors burst, we will need to change interface. But it can be done in following patches